KAFKA-6166: Streams configuration requires consumer. and producer. in order to be read

h314to opened the pull request:
* Implement method to get custom properties
* Add custom properties to getConsumerConfigs and getProducerConfigs
* Add tests

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

h314to committed:
KAFKA-6166: Streams configuration requires consumer. and producer. in order to be read

* Implement method to get custom properties
* Add custom properties to getConsumerConfigs and getProducerConfigs
* Add tests
* Add custom properties to getAdminConfigs
* Address reviewer comments.
* Remove redundant tests
* Remove unnecessary parameters from getClientCustomProps

h314to commented:
retest this please

h314to commented:
The error with JDK8 and Scala 2.12 seems unrelated to any changes introduced in the code:

`02:04:20 ERROR: No tool found matching GRADLE_3_4_RC_2_HOME
02:04:20 Setting GRADLE_3_5_HOME=/home/jenkins/tools/gradle/3.5
02:04:21  FAILED`

Should I rerun the tests?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -687,12 +687,17 @@ public StreamsConfig(final Map<?, ?> props) {
 
     private Map<String, Object> getCommonConsumerConfigs() {
         final Map<String, Object> clientProvidedProps = getClientPropsWithPrefix(CONSUMER_PREFIX, ConsumerConfig.configNames());
+        final Map<String, Object> clientProvidedCustomProps = getClientCustomProps(CONSUMER_PREFIX, PRODUCER_PREFIX);
```
I think, we need also remove AdminClient configs, too.

h314to replied:
Yes, I forgot those. Thanks.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -687,12 +687,17 @@ public StreamsConfig(final Map<?, ?> props) {
         checkIfUnexpectedUserSpecifiedConsumerConfig(clientProvidedProps, NON_CONFIGURABLE_CONSUMER_EOS_CONFIGS);
 
+        checkIfUnexpectedUserSpecifiedConsumerConfig(clientProvidedCustomProps, NON_CONFIGURABLE_CONSUMER_DEFAULT_CONFIGS);
+        checkIfUnexpectedUserSpecifiedConsumerConfig(clientProvidedCustomProps, NON_CONFIGURABLE_CONSUMER_EOS_CONFIGS);
```
Why do we need this? For custom configs, we don't enforce any restrictions.

h314to replied:
I wanted to make sure no non-configurable properties sneaked through via the custom properties. But you are right, since the consumer/producer properties are removed form the custom properties, this is redundant.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -687,12 +687,17 @@ public StreamsConfig(final Map<?, ?> props) {
+
         final Map<String, Object> consumerProps = new HashMap<>(eosEnabled ? CONSUMER_EOS_OVERRIDES : CONSUMER_DEFAULT_OVERRIDES);
         consumerProps.putAll(clientProvidedProps);
+        consumerProps.putAll(clientProvidedCustomProps);
```
I think, we should add those before the other, to make sure that `consumer.myConfig` overwrites `myConfig`. We should add a test for this case, too.

h314to replied:
Nice one. Thanks. I'll change the order and add a test for that.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -820,12 +825,15 @@ private void checkIfUnexpectedUserSpecifiedConsumerConfig(final Map<String, Obje
+        final Map<String, Object> clientProvidedCustomProps = getClientCustomProps(CONSUMER_PREFIX, PRODUCER_PREFIX);
 
         checkIfUnexpectedUserSpecifiedConsumerConfig(clientProvidedProps, NON_CONFIGURABLE_PRODUCER_EOS_CONFIGS);
+        checkIfUnexpectedUserSpecifiedConsumerConfig(clientProvidedCustomProps, NON_CONFIGURABLE_PRODUCER_EOS_CONFIGS);
```
as above

h314to replied:
OK.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -820,12 +825,15 @@ private void checkIfUnexpectedUserSpecifiedConsumerConfig(final Map<String, Obje
         // generate producer configs from original properties and overridden maps
         final Map<String, Object> props = new HashMap<>(eosEnabled ? PRODUCER_EOS_OVERRIDES : PRODUCER_DEFAULT_OVERRIDES);
         props.putAll(clientProvidedProps);
+        props.putAll(clientProvidedCustomProps);
```
as above

h314to replied:
OK.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -855,6 +863,26 @@ private void checkIfUnexpectedUserSpecifiedConsumerConfig(final Map<String, Obje
+        props.keySet().removeAll(ConsumerConfig.configNames());
+        props.keySet().removeAll(ProducerConfig.configNames());
+        props.keySet().removeAll(originalsWithPrefix(consumerPrefix, false).keySet());
+        props.keySet().removeAll(originalsWithPrefix(producerPrefix, false).keySet());
```
as above: we need to remove adminPrefix configs

h314to replied:
Yes, I'll do that.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -264,6 +264,72 @@ public void shouldSupportNonPrefixedProducerConfigs() {
+        final StreamsConfig streamsConfig = new StreamsConfig(props);
+        props.put("custom.property.host", "host");
+        final Map<String, Object> consumerConfigs = streamsConfig.getConsumerConfigs("groupId", "clientId");
+        assertEquals("host", consumerConfigs.get("custom.property.host"));
```
this seems to be equivalent to the "DUMMY" we have above -- I would suggest to remove "DUMMY" from default config as this (and below ones) add proper testing.

I think we can merge this three test and rename `shouldForwardCustomConfigsWithNoPrefixToAllClients`.

We should also add a check for `AdmitClientConfig`.

h314to replied:
Yes, I agree. That will make it cleaner. I'll change it.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -264,6 +264,72 @@ public void shouldSupportNonPrefixedProducerConfigs() {
+        final Map<String, Object> consumerConfigs = streamsConfig.getConsumerConfigs("groupId", "clientId");
+        final Map<String, Object> producerConfigs = streamsConfig.getProducerConfigs("clientId");
+        assertEquals("host", consumerConfigs.get("custom.property.host"));
+        assertNull(producerConfigs.get("custom.property.host"));
```
add check for restore-consumer and admitclient

h314to replied:
Ok

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -264,6 +264,72 @@ public void shouldSupportNonPrefixedProducerConfigs() {
+    }
+
+    @Test
+    public void shouldAddCustomNonPrefixedPropertiesToProducerAndConsumerConfig() {
```
This seems to be redundant from above (or already the merged test...) -- restore-consumer and adminclient would be missing though.

h314to replied:
Yes, I'll clean it up.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -264,6 +264,72 @@ public void shouldSupportNonPrefixedProducerConfigs() {
+        final Map<String, Object> consumerConfigs = streamsConfig.getConsumerConfigs("groupId", "clientId");
+        final Map<String, Object> producerConfigs = streamsConfig.getProducerConfigs("clientId");
+        assertNull(consumerConfigs.get("custom.property.host"));
+        assertEquals("host", producerConfigs.get("custom.property.host"));
```
as above -- add check for two missing clients

h314to replied:
Ok

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -264,6 +264,72 @@ public void shouldSupportNonPrefixedProducerConfigs() {
+    }
+
+    @Test
+    public void shouldWorkIfThereAreNoCustomConfigs() {
```
seems to be covered by all other tests?

h314to replied:
I wanted to make sure the custom properties code wasn't breaking anything (e.g. when there were no custom configs), and this wasn't being tested because the properties used in this class contained that "dummy" custom property. Since I'm removing that property this doesn't do anything anymore. I'll remove it.

mjsax reviewed (changes requested):
Thanks for the PR. Couple of comments.

h314to reviewed (commented):
Thanks for the review. I'll make the suggested changes.

h314to commented:
In the updated PR, apart from addressing the review suggestions, I'm also adding the custom properties to the AdminClient config. I did so to make sure all clients behave in the same way (i.e., custom properties are passed to all). Please confirm that this is ok.

h314to commented:
The 8 tests from `org.apache.kafka.streams.integration.ResetIntegrationTest` failing on all builds  have, according to Jenkins, also been failing for previous PR. Moreover, they are all passing locally, so I don't think the failures are caused by the code in this PR.

h314to commented:
rebased on trunk to get the fixes on `KAFKA-6462: fix unstable ResetIntegrationTest `

guozhangwang commented:
retest this please

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -264,6 +261,79 @@ public void shouldSupportNonPrefixedProducerConfigs() {
+        props.put("custom.property.host", "host");
+        props.put(consumerPrefix("custom.property.host"), "localhost");
+        props.put(producerPrefix("custom.property.host"), "localhost");
+        props.put(adminClientPrefix("custom.property.host"), "localhost");
```
I think, we should use three different values to make sure that the different prefixes overwrite the configs for the corresponding clients. 

Looking into the test below, they seem to be redundant with this one? We can also remove this test and keep the other three (that would avoid redundancy, too)

h314to replied:
Yes, using three different values is definitely safer. In addition I think that actually makes the other three tests redundant because it both checks that prefixed properties override non-prefixed ones, and also that prefixed properties are only sent to the specified client. I'm therefore changing this test as suggested and removing the other three. Let me know if you think I should keep them instead.

mjsax reviewed (commented):
Two more nits. Otherwise LGTM.

guozhangwang commented:
retest this please

guozhangwang commented:
@mjsax not sure if the RestIntegrationTest can still hang after you commit, please ping me if you think so and I will have another look.

mjsax commented:
Not sure. Was this rebased to incorporate the fix?

mjsax reviewed (approved):
LGTM

mjsax commented:
Retest this please.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -862,6 +871,31 @@ private void checkIfUnexpectedUserSpecifiedConsumerConfig(final Map<String, Obje
+     * @param adminClientPrefix the prefix used to denote AdminClient-only properties
+     * @return a map with the custom properties
+     */
+    private Map<String, Object> getClientCustomProps(final String consumerPrefix,
```
Do we need to pass in these three parameters if they are supposed to never change and all its callers are passing the same values?

h314to replied:
Thanks for the review! Yes, definitely unnecessary. I'm removing the arguments. Additionally, since we're just adding the custom properties to the  props map, I'm moving the call to `props.putAll(getClientCustomProps())` instead assigning it to a variable.

guozhangwang reviewed (commented):
Just one nit comment, otherwise LGTM. Thanks @h314to !

guozhangwang commented:
LGTM.

guozhangwang commented:
Also cherry-picked to 1.0 branch so the up coming bug fix release 1.0.1 would also have this.

Thanks @h314to for your contribution and your patience!

h314to commented:
Great! Thank you all for your help reviewing it and making it better.
